src/Icicle/Avalanche/Statement/Simp/Mutate.hs

```diff
@@ -0,0 +1,174 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE PatternGuards #-}
+module Icicle.Avalanche.Statement.Simp.Mutate where
+
+import           Icicle.Common.Exp
+import           Icicle.Common.Base
+import qualified Icicle.Avalanche.Prim.Flat     as F
+import           Icicle.Avalanche.Statement.Statement
+
+import P
+
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           Data.Hashable
+
+
+mutate
+  :: (Hashable n, Eq n)
+  => Statement a n F.Prim
+  -> Statement a n F.Prim
+mutate = mutateStmt Set.empty
+
+
+mutateStmt
+  :: (Hashable n, Eq n)
+  => Set (Name n)
+  -> Statement a n F.Prim -> Statement a n F.Prim
+mutateStmt livevars ss
+  = let go = mutateStmt livevars
+    in case ss of
+         If x s1 s2
+           -> let lvs = livevars <> livevarsStmt s1 <> livevarsStmt s2
+              in If (mutateExp lvs x)
+                    (go s1)
+                    (go s2)
+
+         Let n x s
+           -> let lvs = livevars <> livevarsStmt s
+              in  Let n (mutateExp lvs x)
+                        (go s)
+
+         ForeachInts n x y s
+           -> let lvs = livevars <> livevarsStmt s
+              in  ForeachInts n (mutateExp livevars x)
+                                (mutateExp lvs y)
+                                (go s)
+
+         ForeachFacts fs vt ft s
+           -> ForeachFacts fs vt ft (go s)
+
+         Block stmts
+           -> let f (st:sts) = mutateStmt (livevars <> Set.unions (fmap livevarsStmt sts)) st
+                              : fmap go sts
```

Comment:
I think the first part here is right, getting the livevars of all the remaining sts, but shouldn't this be `f sts` rather than `fmap go sts`? as it is, each successive substatement will lose the livevars of those after it